Cutelyst/context.cpp

```diff
@@ -499,23 +499,27 @@ void Context::finalize()
     Q_D(Context);
 
     if (d->stats) {
-        qCDebug(CUTELYST_STATS, "Response Code: %d; Content-Type: %s; Content-Length: %s",
-                d->response->status(),
-                qPrintable(d->response->headers().header(QStringLiteral("CONTENT_TYPE"), QStringLiteral("unknown"))),
-                qPrintable(d->response->headers().header(QStringLiteral("CONTENT_LENGTH"), QStringLiteral("unknown"))));
-
-        const double enlapsed = d->engineRequest->elapsed.nsecsElapsed() / 1000000000.0;
-        QString average;
-        if (enlapsed == 0.0) {
-            average = QStringLiteral("??");
-        } else {
-            average = QString::number(1.0 / enlapsed, 'f');
-            average.truncate(average.size() - 3);
+        // statistics collection is enabled
+        // either by enabling logging category cutelyst.stats.debug=true
+        // or by using the Prometheus plugin
+        if (CUTELYST_STATS().isDebugEnabled()) {
```

Comment:
This is not needed as d->stats is previously calculated by it

Reply:
Sorry, but I don't understand your comment. The purpose of the line you cite is to suppress debug output, if the Prometheus plugin enables the statistics. Only if the user requests debug output via cutelyst.stats.debug=true, the output is generated. The only change I introduce here is the if-statement. Everything else is your original code with new indent.
Can you please explain the problem in more detail?

Reply:
I think this:
https://github.com/cutelyst/cutelyst/blob/master/Cutelyst/application.cpp#L258
and
https://github.com/cutelyst/cutelyst/blob/master/Cutelyst/application.cpp#L377
will clarify what I am talking about.

Reply:
The program flow is like this:
1. https://github.com/cutelyst/cutelyst/blob/fceca650d1f52f6982618daa63a03c1aa95a8b35/Cutelyst/application.cpp#L258
1. (cannot link to it :worried:) prometheus.cpp#L67
So, stats are enabled unconditionally. I cannot use d->useStats to decide if debug messages should be printed to the console.
If I understand correctly, I should leave the file context.cpp alone (revert my changes) and let the application calculate the values `elapsed` and `average`, even if they are not printed to the console, because debug output is off.

I'll change that in my next revision of the pull request, but I identified one issue while stepping through the code, that needs to be fixed. Hope to get it fixed soon.

Reply:
hmm I didn't notice you were doing that private useStats change, that's not quite good.
Also this if CUTELYST_STATS().isDebugEnabled() is bit costly which is why I don't want it to be checked all the time, I know there are other places that also have this, but the less the better.

IMO what we need is something that I was supposed to do some time ago, that is have a way to have an user supplied Stats class, that way you could add your class and be done. But that isn't much easy work especially due it being instantiated on every request..

Reply:
Instrumenting an application is always costly... The evaluation of CUTELYST_STATS().isDebugEnabled() is only done, if statistics are enabled in the first place.
I've no problem postponing this pull request, until you have the statistics feature implemented.
I'll need this plugin for myself to monitor the performance of my app and will therefore use it as a patchset on top of mainline. I'll push any improvement to the plugin here; can we set this pull request to draft again?